### PR TITLE
 parser: deny all rust keywords as target names  and add `gen` to keywords

### PR DESCRIPTION
**Pull request by Kijewski**
Resolves https://github.com/rinja-rs/askama/issues/1048.

(Does this work across repos?)

### CODE REVIEW

File: rinja_parser/src/lib.rs [hunk 1]
```diff
@@ -1122,13 +1122,123 @@
     list
 };
 
+#[doc(hidden)]
+pub const MAX_RUST_KEYWORD_LEN: usize = 8;
+const MAX_RUST_RAW_KEYWORD_LEN: usize = MAX_RUST_KEYWORD_LEN + 2;
+#[doc(hidden)]
+pub const RUST_KEYWORDS: &[&[[u8; MAX_RUST_RAW_KEYWORD_LEN]]] = {
+    // FIXME: Replace `u8` with `[core:ascii::Char; MAX_REPL_LEN]` once
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply rinja_parser/src/lib.rs:1130]: Not a new message, just moved.

File: rinja_parser/src/node.rs [hunk 2]
```diff
@@ -1161,6 +1161,31 @@ impl<'a> Let<'a> {
...
+                | Target::BoolLit(..) => Some("a literal"),
+                Target::Path(..) => Some("a path or enum variant"),
+                Target::OrChain(..) | Target::Placeholder(..) | Target::Rest(..) => {
+                    Some("a pattern")
+                }
+            };
+            if let Some(kind) = kind {
+                return Err(winnow::error::ErrMode::Cut(ErrorContext::new(
+                    format!(
+                        "when you forward-define a variable, you cannot use {kind} inplace of \
```
Comment (GuillaumeGomez) [inline rinja_parser/src/node.rs:1182]:
```suggestion
                        "when you forward-define a variable, you cannot use {kind} in place of \
```

File: rinja_parser/src/target.rs [hunk 3]
```diff
@@ -194,7 +195,7 @@ fn verify_name<'a>(
     input: &'a str,
     name: &'a str,
 ) -> Result<Target<'a>, winnow::error::ErrMode<ErrorContext<'a>>> {
-    if name == "self" {
+    if is_rust_keyword(name) {
         Err(winnow::error::ErrMode::Cut(ErrorContext::new(
             format!("cannot use `{name}` as a name"),
```
Comment (GuillaumeGomez) [inline rinja_parser/src/target.rs:200]: I would update the error message to be:

```
"cannot use `{name}` as a variable name: it is a rust keyword"
```

File: rinja_parser/src/node.rs (hunk 2 above)
Comment (Kijewski) [reply rinja_parser/src/node.rs]: The dictionary says you're right: <https://www.merriam-webster.com/dictionary/place#phrases-anchor>. Thank you for proofreading! :)

File: rinja_parser/src/target.rs (hunk 3 above)
Comment (Kijewski) [reply rinja_parser/src/target.rs]: Good idea! Changed.

File: rinja_parser/src/target.rs [hunk 4]
```diff
@@ -194,14 +195,14 @@ fn verify_name<'a>(
...
+    if is_rust_keyword(name) {
         Err(winnow::error::ErrMode::Cut(ErrorContext::new(
-            format!("cannot use `{name}` as a name"),
+            format!("cannot use `{name}` as a name: it is a rust keyword"),
             input,
         )))
     } else if name.starts_with("__rinja") {
         Err(winnow::error::ErrMode::Cut(ErrorContext::new(
-            format!("cannot use `{name}` as a name: it's reserved for `rinja`"),
+            format!("cannot use `{name}` as a name: it is reserved for `rinja`"),
```
Comment (GuillaumeGomez) [inline rinja_parser/src/target.rs:205]: Ah nice improvement. Didn't even think about it. ^^'

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks!

**Pull request merged**